types/wlr_xwayland_keyboard_grab_v1.c

```diff
@@ -0,0 +1,200 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <wayland-server.h>
+#include <wayland-util.h>
+#include <wlr/util/log.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_xwayland_keyboard_grab_v1.h>
+
+#include <xwayland-keyboard-grab-unstable-v1-protocol.h>
+
+struct wlr_xwayland_keyboard_grab_v1_manager {
+	struct wl_resource *resource;
+	struct wlr_xwayland_keyboard_grab_v1 *xwayland_keyboard_grab;
+
+	struct wl_list link; // wlr_xwayland_keyboard_grab_v1::clients;
+	struct wl_list grabs;
+
+	struct {
+		struct wl_signal destroy;
+	} events;
+};
+
+static void xwayland_keyboard_grab_grab_handle_destroy(struct wl_resource *resource) {
+	assert(resource);
+	struct wlr_xwayland_keyboard_grab_v1_grab *grab =
+		wl_resource_get_user_data(resource);
+	assert(grab);
+
+	wl_signal_emit(&grab->events.destroy, grab);
+	wl_list_remove(&grab->link);
+	free(grab);
+}
+
+static void xwayland_keyboard_grab_grab_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct zwp_xwayland_keyboard_grab_v1_interface xwayland_keyboard_grab_impl = {
+	.destroy = xwayland_keyboard_grab_grab_destroy,
+};
+
+static void xwayland_keyboard_grab_grab_keyboard(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id,
+		struct wl_resource *surface, struct wl_resource *seat) {
+	assert(client);
+	assert(resource);
+	assert(surface);
+	assert(seat);
```

Comment:
All these `assert` are useless, Wayland ensures these aren't NULL (there's a protocol property to allow null values).